lib/jxl/splines.cc

```diff
@@ -507,10 +549,13 @@ Status Splines::InitializeDrawCache(size_t image_xsize, size_t image_ysize,
   segment_indices_.clear();
   segment_y_start_.clear();
   std::vector<std::pair<size_t, size_t>> segments_by_y;
+  Spline spline;
+  // TODO(eustas): not in the spec; limit splines feature up to 32MPx.
+  size_t px_limit = 1 << 25;
```

Comment:
This limits the total rendered spline area to 32 megapixels, right?

That might be a somewhat limiting factor for jxl art :)
(e.g. you can have a 4 megapixel image with a bunch of overlapping big splines that could together maybe have a nominal area larger than 32 megapixels)

It makes sense to limit the total area though - that's the main thing that will determine memory and cpu cost to decode.

But maybe we should put a limit on it that is a bit less strict and maybe also make it depend on the image dimensions (a 100x100 image with 32MPx worth of splines is something else than a 10000x10000 images with 32 MPx worth of splines, imo).

And we should put this limit in the spec then.

I would suggest to limit the total spline area to `(1 << 16) + (xsize * ysize << 4)` or something like that (i.e. allow every pixel to be part of at most 16 overlapping splines on average).

Opinions welcome. Ping @veluca93 @lvandeve @lovell

Reply:
I like the idea of an adaptive limit

Reply:
I propose just to have some preliminary limit at the moment.
Until splines are rendered group-wise size-based limit could become a CPU/memory threat.
Also not sure that this should be specced. Spline rendering should allow some quality trade-offs.

Reply:
I'm not sold of that - we have plenty of things that are not rendered group-wise and use O(image) memory, so I don't see the issue of a per-image-size limit.

Reply:
Quality trade-offs is one thing. Refusing to decode an image is something else. The whole point of profiles and levels is to establish some subset of the bitstream that is guaranteed to decode. If we start introducing de facto extra limits (that are actually limiting, not checks like coords being within max image dimensions range), I think it's better to be explicit about it and add it to the spec.

We can have a low limit and/or one that is not image size dependent, or one that guarantees that groups can be decoded with constant memory (so a limit of spline area per group, not per image). But if it's a limiting limit (e.g. for someone who would want to make an svg to jxl splines converter), we have to make sure the limits are well documented and described in the spec. Otherwise we're basically making a reference implementation that doesn't conform to the spec.

Reply:
Any convergence on this?

Reply:
It looks OK to me as it is now. We'll have to see if and what exactly we put in the spec as a limitation, but for now these limits look sane to me without being restrictive.